components/script/dom/document.rs

```diff
@@ -2708,26 +2692,15 @@ impl Document {
         );
     }
 
-    pub(crate) fn add_dirty_webgl_canvas(&self, context: &WebGLRenderingContext) {
-        self.dirty_webgl_contexts
-            .borrow_mut()
-            .entry(context.context_id())
-            .or_insert_with(|| Dom::from_ref(context));
-    }
-
-    pub(crate) fn add_dirty_2d_canvas(&self, context: &CanvasRenderingContext2D) {
-        self.dirty_2d_contexts
-            .borrow_mut()
-            .entry(context.context_id())
-            .or_insert_with(|| Dom::from_ref(context));
-    }
-
-    #[cfg(feature = "webgpu")]
-    pub(crate) fn add_dirty_webgpu_context(&self, context: &GPUCanvasContext) {
-        self.dirty_webgpu_contexts
-            .borrow_mut()
-            .entry(context.context_id())
-            .or_insert_with(|| Dom::from_ref(context));
+    pub(crate) fn mark_canvas_as_dirty(&self, canvas: &HTMLCanvasElement) {
+        let mut dirty_canvases = self.dirty_canvases.borrow_mut();
+        if dirty_canvases
+            .iter()
+            .any(|dirty_canvas| std::ptr::eq(&**dirty_canvas, canvas))
```

Comment:
If we import the `DomObject` trait, we can use the `PartialEq` of dom structs to do `dirty_canvas.reflector() == canvas.reflector()`.

Or even better, since we create `Dom::ref(canvas)` in the last line of the method, we can move it up before this line and just do `*dirty_canvas == canvas`.